File: Samples/Java/SpringBoot/src/main/java/io/cratis/arc/samples/javaspringboot/CreateTaskValidator.java
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

package io.cratis.arc.samples.javaspringboot;

import io.cratis.arc.commands.CommandContext;
import io.cratis.arc.java.BlockingCommandValidator;
import io.cratis.arc.results.ValidationResult;
import io.cratis.arc.results.ValidationResultSeverity;
import java.util.List;

/** Validates task creation requests through Arc's public validation seam. */
public final class CreateTaskValidator implements BlockingCommandValidator<CreateTask> {
    private static final int MAXIMUM_TITLE_LENGTH = 80;

    @Override
    public Class<CreateTask> getCommandType() {
        return CreateTask.class;
    }

    @Override
    public List<ValidationResult> validate(CreateTask command, CommandContext context) {
        if (command.title() == null || command.title().isBlank()) {
            return List.of(new ValidationResult(
                ValidationResultSeverity.Error,
                "A task title is required.",
                List.of("title")));
        }
        if (command.title().length() > MAXIMUM_TITLE_LENGTH) {
            return List.of(new ValidationResult(
                ValidationResultSeverity.Error,
                "A task title cannot exceed " + MAXIMUM_TITLE_LENGTH + " characters.",
                List.of("title")));
        }
        return List.of();
    }
}

File: Samples/Kotlin/SpringBoot/src/main/kotlin/io/cratis/arc/samples/kotlin/springboot/CreateTaskValidator.kt
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

package io.cratis.arc.samples.kotlin.springboot

import io.cratis.arc.commands.CommandContext
import io.cratis.arc.commands.CommandValidator
import io.cratis.arc.results.ValidationResult
import io.cratis.arc.results.ValidationResultSeverity
import org.springframework.stereotype.Component

/** Validates task creation requests through Arc's public validation seam. */
@Component
public class CreateTaskValidator : CommandValidator<CreateTask> {
    override val commandType: Class<CreateTask> = CreateTask::class.java

    override suspend fun validate(command: CreateTask, context: CommandContext): List<ValidationResult> = when {
        TaskTitleRule.isBlank(command.title) -> listOf(
            ValidationResult(ValidationResultSeverity.Error, "A task title is required.", listOf("title"))
        )
        TaskTitleRule.isTooLong(command.title) -> listOf(
            ValidationResult(
                ValidationResultSeverity.Error,
                "A task title cannot exceed ${TaskTitleRule.MAXIMUM_LENGTH} characters.",
                listOf("title")
            )
        )
        else -> emptyList()
    }
}
